pkg/jobstore/boltdb/store.go

```diff
@@ -0,0 +1,1013 @@
+	}
+
+	jobIDKey := []byte(job.Metadata.ID)
+
+	jobState := model.JobState{
+		JobID:      job.Metadata.ID,
+		State:      model.JobStateNew,
+		Version:    1,
+		CreateTime: b.clock.Now().UTC(),
+		UpdateTime: b.clock.Now().UTC(),
+	}
+	data, err := json.Marshal(jobState)
+	if err != nil {
+		return err
+	}
+
+	// Write the JobState to the state bucket
+	if bkt, err := NewBucketPath(BucketPathState).Get(tx, false); err != nil {
+		return err
+	} else {
+		if err = bkt.Put(jobIDKey, data); err != nil {
+			return err
+		}
+	}
+
+	// Write the job to the Job bucket
+	jobData, err := json.Marshal(job)
+	if err != nil {
+		return err
+	}
+	if bkt, err := NewBucketPath(BucketPathJobs).Get(tx, false); err != nil {
+		return err
+	} else {
+		if err = bkt.Put(jobIDKey, jobData); err != nil {
+			return err
+		}
+	}
+
+	if err = b.inProgressIndex.Add(tx, BucketJobsInProgress, jobIDKey); err != nil {
+		return err
+	}
+
+	if job.Metadata.ClientID == "" {
+		fmt.Println("TEST missing clientID in job")
+		return errors.New("job is missing a client id")
+	}
+
+	if err = b.clientsIndex.Add(tx, []byte(job.Metadata.ClientID), jobIDKey); err != nil {
+		return err
+	}
+
+	// Write sentinels keys for specific tags
+	for _, tag := range job.Spec.Annotations {
+		tagBytes := []byte(strings.ToLower(tag))
+		if err = b.tagsIndex.Add(tx, tagBytes, jobIDKey); err != nil {
+			return err
+		}
+	}
+
+	b.triggerEvent(jobstore.JobWatcher, jobstore.CreateEvent, jobData)
+
+	return b.appendJobHistory(tx, jobState, model.JobStateNew, newJobComment)
+}
+
+// DeleteJob removes the specified job from the system entirely
+func (b *BoltJobStore) DeleteJob(ctx context.Context, jobID string) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.deleteJob(tx, jobID)
+	})
+}
+
+func (b *BoltJobStore) deleteJob(tx *bolt.Tx, jobID string) error {
+	jobIDKey := []byte(jobID)
+
+	job, err := b.getJob(tx, jobID)
+	if err != nil {
+		return bacerrors.NewJobNotFound(jobID)
+	}
+
+	// Delete the JobState from the state bucket
+	if bkt, err := NewBucketPath(BucketPathState).Get(tx, false); err != nil {
+		return err
+	} else {
+		if err = bkt.Delete(jobIDKey); err != nil {
+			return err
+		}
+	}
+
+	// Delete the actual job
+	if bkt, err := NewBucketPath(BucketPathJobs).Get(tx, false); err != nil {
+		return err
+	} else {
+		if err = bkt.Delete(jobIDKey); err != nil {
+			return err
+		}
+	}
+
+	if err = b.inProgressIndex.Remove(tx, BucketJobsInProgress, jobIDKey); err != nil {
+		return err
+	}
+
+	if err = b.clientsIndex.Remove(tx, []byte(job.Metadata.ClientID), jobIDKey); err != nil {
+		return err
+	}
+
+	// Delete sentinels keys for specific tags
+	for _, tag := range job.Spec.Annotations {
+		tagBytes := []byte(strings.ToLower(tag))
+		if err = b.tagsIndex.Remove(tx, tagBytes, jobIDKey); err != nil {
+			return err
+		}
+	}
+
+	encodedJob, _ := json.Marshal(job)
+	b.triggerEvent(jobstore.JobWatcher, jobstore.DeleteEvent, encodedJob)
```

Comment:
wouldn't it be better to pass the job object instead of its bytes in the event?

Reply:
Agreed